### PR TITLE
[SLP]Improve masked loads vectorization, attempting gathered loads

### PR DESCRIPTION
**Pull request by alexey-bataev**
If the vector of loads can be vectorized as masked gather and there are
several other masked gather nodes, compiler can try to attempt to check,
if it possible to gather such nodes into big consecutive/strided loads
  node, which provide better performance.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -5118,10 +5125,10 @@ BoUpSLP::canVectorizeLoads(ArrayRef<Value *> VL, const Value *VL0,
       L && Sz > 2 && static_cast<unsigned>(count_if(PointerOps, [L](Value *V) {
                        return L->isLoopInvariant(V);
                      })) <= Sz / 2;
-  if (ProfitableGatherPointers || all_of(PointerOps, [IsSorted](Value *P) {
+  if (ProfitableGatherPointers || all_of(PointerOps, [](Value *P) {
         auto *GEP = dyn_cast<GetElementPtrInst>(P);
-        return (IsSorted && !GEP && doesNotNeedToBeScheduled(P)) ||
-               (GEP && GEP->getNumOperands() == 2 &&
+        return (!GEP && doesNotNeedToBeScheduled(P)) ||
+               (GEP->getNumOperands() == 2 &&
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5131]: Can GEP not still be NULL here?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -3614,6 +3616,13 @@ class BoUpSLP {
       DenseMap<Value *, SmallPtrSet<const TreeEntry *, 4>>;
   ValueToGatherNodesMap ValueToGatherNodes;
 
+  /// A list of the loads, which can be vectorized using strided or masked
+  /// gather approach, but attempted to be represented as contiguous loads.
+  SetVector<unsigned> LoadEntriesToVectorize;
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3621]: What does the unsigned value represent?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -3614,6 +3616,13 @@ class BoUpSLP {
...
+
+  /// true if graph nodes transforming mode is on.
+  bool IsGraphTransformMode = false;
+
   /// The index of the first gathered load entry in the VectorizeTree.
   constexpr static int NoGatheredLoads = -1;
   int GatheredLoadsEntriesFirst = NoGatheredLoads;
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3628]: (off topic) - Could this be `std::optional<unsigned> GatheredLoadsEntriesFirst = None` ?

File: llvm/test/Transforms/SLPVectorizer/RISCV/remarks-insert-into-small-vector.ll [hunk 4]
```diff
@@ -8,20 +8,17 @@
...
 ; CHECK-LABEL: define void @test(
 ; CHECK-SAME: ) #[[ATTR0:[0-9]+]] {
 ; CHECK-NEXT:  entry:
-; CHECK-NEXT:    [[TMP0:%.*]] = load float, ptr null, align 4
-; CHECK-NEXT:    [[TMP1:%.*]] = load float, ptr null, align 4
 ; CHECK-NEXT:    [[TMP2:%.*]] = load float, ptr null, align 4
-; CHECK-NEXT:    [[TMP3:%.*]] = insertelement <2 x float> <float poison, float 0.000000e+00>, float [[TMP1]], i32 0
-; CHECK-NEXT:    [[TMP4:%.*]] = insertelement <2 x float> poison, float [[TMP0]], i32 0
-; CHECK-NEXT:    [[TMP5:%.*]] = insertelement <2 x float> [[TMP4]], float [[TMP2]], i32 1
+; CHECK-NEXT:    [[TMP5:%.*]] = call <2 x float> @llvm.masked.gather.v2f32.v2p0(<2 x ptr> zeroinitializer, i32 4, <2 x i1> <i1 true, i1 true>, <2 x float> poison)
```
Comment (RKSimon) [inline llvm/test/Transforms/SLPVectorizer/RISCV/remarks-insert-into-small-vector.ll:20]: Why did this change? Its not sequential

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 1 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:5131]: Added a check

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 2 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3621]: Fixed comment, describing the container

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3628]: Yes, if you prefer this style. Can do in a separate patch

File: llvm/test/Transforms/SLPVectorizer/RISCV/remarks-insert-into-small-vector.ll (hunk 4 above)
Comment (alexey-bataev) [reply llvm/test/Transforms/SLPVectorizer/RISCV/remarks-insert-into-small-vector.ll:20]: It is still not sequential, just a side effect of a less restrictive check for ProfitableGatherPointers

**Review by RKSimon — APPROVED** [review]

**Pull request merged**